examples/tutorials/core_api_mnist/model_def_adaptive.py

```diff
@@ -0,0 +1,250 @@
+    )
+    parser.add_argument(
+        "--save-model", action="store_true", default=True, help="For Saving the current Model"
+    )
+
+    args = parser.parse_args()
+    use_cuda = not args.no_cuda and torch.cuda.is_available()
+    use_mps = not args.no_mps and torch.backends.mps.is_available()
+
+    info = det.get_cluster_info()
+    assert info is not None, "this example only runs on-cluster"
+    latest_checkpoint = info.latest_checkpoint
+    if latest_checkpoint is not None:
+        with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+            model = load_state(path)
+
+    torch.manual_seed(args.seed)
+
+    if use_cuda:
+        device = torch.device("cuda")
+    elif use_mps:
+        device = torch.device("mps")
+    else:
+        device = torch.device("cpu")
+
+    train_kwargs = {"batch_size": args.batch_size}
+    test_kwargs = {"batch_size": args.test_batch_size}
+    if use_cuda:
+        cuda_kwargs = {"num_workers": 1, "pin_memory": True, "shuffle": True}
+        train_kwargs.update(cuda_kwargs)
+        test_kwargs.update(cuda_kwargs)
+
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
+    )
+    dataset1 = datasets.MNIST("../data", train=True, download=True, transform=transform)
+    dataset2 = datasets.MNIST("../data", train=False, transform=transform)
+    train_loader = torch.utils.data.DataLoader(dataset1, **train_kwargs)
+    test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)
+
+    # NEW: Get hparams chosen for this trial from cluster info object
+    hparams = info.trial.hparams
+
+    # NEW: Pass relevant hparams to model and optimizer
+    model = Net(hparams).to(device)
+    optimizer = optim.Adadelta(model.parameters(), lr=hparams["learning_rate"])
+
+    scheduler = StepLR(optimizer, step_size=1, gamma=args.gamma)
+
+    # NEW: Iterate through the core_context.searcher.operations() to decide how long to train for.
+    starting_epoch = 0
+    epoch = starting_epoch
+    last_checkpoint_batch = None
+
+    for op in core_context.searcher.operations():
+
+        # NEW: Use a while loop for easier accounting of absolute lengths.
+        while epoch < op.length:
+            steps_completed = epoch * len(train_loader)
+
+            # NEW: Pass op into train and test functions
+            train(args, model, device, train_loader, optimizer, core_context, epoch, op)
+            test_loss = test(args, model, device, test_loader, core_context, steps_completed, op)
+
+            scheduler.step()
+            if args.save_model:
+
+                checkpoint_metadata_dict = {
+                    "steps_completed": steps_completed,
+                }
+
+            epoch += 1
+
+            with core_context.checkpoint.store_path(checkpoint_metadata_dict) as (path, storage_id):
+                torch.save(model.state_dict(), str(path) + ("/checkpoint.pt"))
```

Comment:
you changed this line from how it was done in `model_def_checkpoints.py` but it was cleaner in that version.